packages/uploadthing/src/internal/handler.ts

```diff
@@ -1,649 +1,543 @@
-      case "upload": {
-        const maybeInput = await safeParseJSON<UTEvents["upload"]["in"]>(req);
+      if (hook === "callback") {
+        // This is when we receive the webhook from uploadthing
+        return yield* $(
+          handleCallbackRequest({
+            req: request,
+            uploadable,
+          }),
+        );
+      }
 
-        if (maybeInput instanceof Error) {
-          logger.error("Invalid request body", maybeInput);
-          return new UploadThingError({
-            code: "BAD_REQUEST",
-            message: "Invalid request body",
-            cause: maybeInput,
-          });
+      switch (action) {
+        case "upload": {
+          return yield* $(
+            handleUploadAction({
+              req: request,
+              middlewareArgs: input.middlewareArgs,
+              uploadable,
+              config: opts.config,
+              isDev,
+              slug,
+            }),
+          );
         }
-
-        logger.debug("Handling upload request with input:", maybeInput);
-        const { files, input: userInput } = maybeInput;
-
-        // Validate without Zod (for now)
-        if (
-          !Array.isArray(files) ||
-          !files.every(
-            (f) =>
-              isObject(f) &&
-              typeof f.name === "string" &&
-              typeof f.size === "number" &&
-              typeof f.type === "string",
-          )
-        ) {
-          const msg = `Expected files to be of type '{name:string, size:number, type:string}[]', got '${JSON.stringify(
-            files,
-          )}'`;
-          logger.error(msg);
-          return new UploadThingError({
-            code: "BAD_REQUEST",
-            message: "Files must be an array of objects with name and size",
-            cause: msg,
-          });
+        case "multipart-complete": {
+          return yield* $(
+            handleMultipartCompleteAction({
+              req: request,
+            }),
+          );
         }
+        case "failure": {
+          return yield* $(
+            handleMultipartFailureAction({
+              req: request,
+              uploadable,
+            }),
+          );
+        }
+        default: {
+          action satisfies never;
+          return yield* $(
+            Effect.fail(
+              new UploadThingError({
+                code: "BAD_REQUEST",
+                message: `Invalid action type ${action as string}`,
+              }),
+            ),
+          );
+        }
+      }
+    }).pipe(
+      // TODO: Maybe look over this error handling
+      // Now we return UploadThingError and die on everything else
+      Effect.catchTag("FetchError", (err) => {
+        logger.error("An error occurred while fetching", err);
+        return Effect.die(err);
+      }),
+      Effect.catchTag("ParseError", (err) => {
+        logger.error("An error occurred while parsing input/output", err);
+        return Effect.die(err);
+      }),
+      Effect.match({
```

Comment:
```ts
Effect.Effect<Value, Error> => Effect.Effect<Value | Error, never>
```

so when we run the handler `await runRequestHandlerAsync()`, we'll get the error out as a value (like we did before) instead of it throwing

Reply:
there might be a built-in that does this though, i haven't dug into all the effect functions out there but there's a bunch of them.